Rework handling of special tokens

francoishernandez opened the pull request:
This PR is an attempt at facilitating configuration of special tokens, and working with some specificities.

Two main changes :
- addition of `{bos,eos,unk,pad}_token` fields in `BaseVocabConfig`, which are then stored in a `"specials"` key in the vocab object;
- addition of `optional_eos` in `PredictConfig` to handle cases where we might need several EOS, e.g. Llama3 with `<|end_of_text|>` and `<|eot_id|>`

Some open questions / TODOs:
- what about the `default_specials` field, we should probably deprecate it in favor of the more flexible new fields ;
- conversion tools would need to be adapted to the new logic;
- do we want to support HF Tokenizers at some point? if not, we need to work on getting pyonmttok upgraded (https://github.com/OpenNMT/Tokenizer/issues/329, https://github.com/OpenNMT/Tokenizer/issues/330)

vince62s commented on eole/predict/inference.py:
```diff
@@ -86,14 +86,25 @@ def __init__(
+        self._tgt_eos_idx = [
+            vocabs["tgt"].lookup_token(vocabs.get("specials", {}).get("eos_token", ""))
+        ] + [vocabs["tgt"].lookup_token(eos_token) for eos_token in optional_eos]
+        print(self._tgt_eos_idx)
```
remove print

vince62s commented on eole/config/run.py:
```diff
@@ -93,6 +93,10 @@ class PredictConfig(
         None  # patch for CT2 inference engine (to improve later)
     )
     model: ModelConfig | None = None
+    optional_eos: List[str] | None = Field(
```
why don't we put this in the same section as bos_token/eos_token ?
It's better within the model I think.

vince62s replied:
second thought should be fine here

vince62s commented on eole/inputters/inputter.py:
```diff
@@ -80,7 +80,12 @@ def _pad_vocab_to_multiple(vocab, multiple):
 
     vocabs["decoder_start_token"] = config.decoder_start_token
-
+    vocabs["specials"] = {
```
I think the order matters. for consistence we need bos=0 pad=1 eos=2 unk=3

vince62s commented on eole/models/model_saver.py:
```diff
@@ -58,6 +59,14 @@ def load_checkpoint(model_path):
                 checkpoint["vocab"] = json.load(f)
+            # use default specials if not specified
+            if "specials" not in checkpoint["vocab"].keys():
+                checkpoint["vocab"]["specials"] = {
```
I think the order matters. for consistence we need bos=0 pad=1 eos=2 unk=3

vince62s commented on eole/predict/prediction.py:
```diff
@@ -42,7 +42,9 @@ def _build_target_tokens(self, src, srclen, pred, attn, voc, dyn_voc):
             ]
-        if tokens[-1] == DefaultTokens.EOS:
+        if tokens[-1] == self.vocabs.get("specials", {}).get(
+            "eos_token", DefaultTokens.EOS
```
only "eos_token" or also `optional_eos` ?

francoishernandez commented:
I think we are mostly good on this one.

In the latest commits I:
- removed the `default_specials` flag (instead, we grab the values from vocabs["specials"] when needed in build_vocab);
- updated `convert_HF` to grab the actual special tokens from `special_tokens_map.json` (might need to extend if alternatives exist on HF, not sure)
- (roughly) updated the FAQ entry;

@vince62s @l-k-11235 feedback welcome!

Note: I removed the [`<0x00>` padding patch](https://github.com/eole-nlp/eole/blob/ff39275c50d12951963008da11d029940b590713/eole/bin/convert/convert_HF.py#L943-L945) in `convert_HF`. I'm not sure if we should 1. keep it 2. adapt it (but in what way?) or 3. ignore it.

Next step should probably be to dive back into #42, and automating the `mapped_tokens`/`optional_eos`/inference config stuff as discussed [here](https://github.com/eole-nlp/eole/discussions/80#discussioncomment-10656670).

vince62s commented:
I think our code requires a padding token

francoishernandez commented:
Indeed, such situations where no padding token is explicitly set are quite dubious. Not sure it would break, because in many cases we fall back to DefaultTokens.PAD or at the very least <unk>, or id 0, but that's relatively unclear.
Also, we still had a lot of cases where DefaultTokens are used instead of the configured specials. Latest commits fix some cases, but stuff like CT2 will require deeper work (if we recover CT2 support at some point).
We'll probably need to test this a bit more in depth, and do another round of cleanup/factorization before merging.

francoishernandez commented:
Let's merge this to move forward.
